### PR TITLE
shrink OpTy back down to 80 bytes

### PR DESCRIPTION
**Pull request by RalfJung**
This shares the first commit with https://github.com/rust-lang/rust/pull/99013, but seeing the perf trouble in that PR, maybe this one is worth considering first. It gets the size of `OpTy` down to how it was before https://github.com/rust-lang/rust/pull/98846.

r? @oli-obk

### CODE REVIEW

**Comment by RalfJung** [comment]
Let's see if this undoes the Max RSS regressions from https://github.com/rust-lang/rust/pull/98846.

@bors try @rust-timer queue

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 1]
```diff
@@ -598,14 +538,20 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
         // Sanity-check the type we ended up with.
-        debug_assert!(mir_assign_valid_types(
-            *self.tcx,
-            self.param_env,
-            self.layout_of(self.subst_from_current_frame_and_normalize_erasing_regions(
-                place.ty(&self.frame().body.local_decls, *self.tcx).ty
-            )?)?,
-            op.layout,
-        ));
+        // Sanity-check the type we ended up with.
```
Comment (fbstj) [inline compiler/rustc_const_eval/src/interpret/operand.rs:553]: duplicate comment?

File: compiler/rustc_const_eval/src/interpret/operand.rs [hunk 2]
```diff
@@ -596,16 +548,22 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/operand.rs:553]:
```suggestion
```

**Comment by RalfJung** [comment]
Uh this is a *regression* for Max RSS? How does that make any sense?

**Comment by RalfJung** [comment]
Blocked on https://github.com/rust-lang/rust/pull/99101.

**Comment by Kobzol** [comment]
I wouldn't worry about the max RSS regression, it has a low significance factor and it's just on a single `hello-world` benchmark, we can safely ignore that.

**Comment by RalfJung** [comment]
All right, finally ready for review. :)

Also just to be sure...
@bors try @rust-timer queue

**Comment by RalfJung** [comment]
I guess that is not too surprising, all these `layout()` calls will not help with perf.

@rust-lang/wg-compiler-performance what do you think? This PR fixes a memory consumption regression (introduced by https://github.com/rust-lang/rust/pull/98846) at the cost of some performance.

**Comment by nnethercote** [comment]
max-rss isn't affected much by this PR. Instructions only regress for `ctfe-stress-5`, which is a stress test and not so important.

But then, I think max-rss also wasn't affected much by #98846, so I'm not sure if this PR is really necessary?

**Comment by RalfJung** [comment]
> But then, I think max-rss also wasn't affected much by https://github.com/rust-lang/rust/pull/98846, so I'm not sure if this PR is really necessary?

Well, I saw the 6% max regression in a primary benchmark and thought that had to be fixed.

Later I learned max-RSS is very noisy, so... if you tell me the RSS numbers in https://github.com/rust-lang/rust/pull/98846 are fine, I am okay with closing this PR.

**Comment by nnethercote** [comment]
I think the RSS numbers in #98846 are fine. For the affected benchmarks, it's only 1 or 2 of the runs, which is a clue that it's just noise. Also, the significance factors are very low, just a little over 1, which is the threshold for displaying. In clearer cases the significant factor will be over 10, sometimes over 50.

All this is unclear with deep understanding of rustc-perf, so apologies for that. We are trying to improve this, but presenting this data in a way that is both precise and comprehensible to non-experts is a genuine challenge.

**Comment by RalfJung** [comment]
> All this is unclear with deep understanding of rustc-perf, so apologies for that. We are trying to improve this, but presenting this data in a way that is both precise and comprehensible to non-experts is a genuine challenge.

I know, and no need to apologize. :) After all, we then have you around to ask questions when things are unclear. :D

I'll close this PR then. I will investigate making `OpTy` not-`Copy` so that we avoid copying such a large type unnecessarily.

**Pull request closed without merging**